AsyncDisplayKit/ASMapNode.mm

```diff
@@ -150,23 +146,33 @@ - (void)takeSnapshot
   }
 }
 
+- (void)setUpSnapshotter
+{
+    if (!_snapshotter) {
+        ASDisplayNodeAssert(!CGSizeEqualToSize(CGSizeZero, self.calculatedSize), @"self.calculatedSize can not be zero. Make sure that you are setting a preferredFrameSize or wrapping ASMapNode in a ASRatioLayoutSpec or similar.");
+        _options.size = self.calculatedSize;
+        _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+    }
+}
+
 - (void)resetSnapshotter
 {
   if (!_snapshotter.isLoading) {
-    _options.size = self.calculatedSize;
     _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
   }
 }
 
-#pragma mark - Action
+#pragma mark - Actions
 - (void)addLiveMap
 {
   if (self.isNodeLoaded && !_mapNode) {
-    _mapNode = [[ASDisplayNode alloc]initWithViewBlock:^UIView *{
-      _mapView = [[MKMapView alloc]initWithFrame:CGRectMake(0.0f, 0.0f, self.calculatedSize.width, self.calculatedSize.height)];
+    __weak ASMapNode *weakSelf = self;
+    _mapNode = [[ASDisplayNode alloc] initWithViewBlock:^UIView *{
+      _mapView = [[MKMapView alloc] initWithFrame:CGRectZero];
       _mapView.delegate = _mapDelegate;
       [_mapView setRegion:_options.region];
```

Comment:
I think there is still a retain cycle here, but probably best to test it with a sample app and Instruments allocation trace to make sure everything gets torn down correctly.  accessing ivars here, every one of them including _options, _mapDelegate, etc will create a strong retain of self.

Also, get rid of the reference to _mapView unless it is needed, and access this through the _mapNode.  I'm a little split on whether we should have _mapNode at all to be honest.

How about we just kill _mapNode and work with _mapView exclusively when dealing with the live map?

In any case, make sure to add ASDisplayNodeAssertMainThread() at the top of this method.